### PR TITLE
#159232837 Disable required fields when creating new asset category

### PR DESCRIPTION
**Pull request by Iyke09**
**What does this PR do?**
It prefills and disables all required fields when creating a new asset category

**Description of Task to be completed?**
- add a disabled option to the required input fields and select box fields
- prepopulate required fields with the required data   

**How should this be manually tested?**
After cloning and running the app, navigate to the settings page and click on add new assets

- you should see the required input fields are disabled and prepopulated with the required data

**Any background context you want to provide?**
N/A

**What are the relevant pivotal tracker stories?**
https://www.pivotaltracker.com/story/show/159232837

**Screenshots (if appropriate)**
screen shot 2018-07-31 at 10 51 04 pm

**Questions:**
N/A

### CODE REVIEW

**Comment by kevinejiro** [comment]
Instead of putting the raw link to your story in your PR, i suggest you follow the teams convention eg [#ID](link to story).

File: src/components/InputBox/index.tsx [hunk 1]
```diff
@@ -7,7 +7,7 @@ import './InputBox.scss';
 // interfaces
 import { InputBoxProps } from './interfaces';
 
-const InputBox = ({ error, placeholder, onChange, value, name }: InputBoxProps) => (
+const InputBox = ({ error, placeholder, onChange, value, name, isDisabled }: InputBoxProps) => (
```
Comment (bahdcoder) [inline src/components/InputBox/index.tsx]: This component is used throughout the project. You should add tests for the `isDisabled` scenarios. This is for security, making sure the component always works as expected.

File: src/pages/CreateCategoryPage/index.tsx [hunk 2]
```diff
@@ -7,21 +7,23 @@ import './CreateCategoryPage.scss';
...
 // interfaces
 import { CreateCategoryPageState } from './interfaces';
 
 class CreateCategoryPage extends React.Component<{}, CreateCategoryPageState> {
+  defaultSelectValues = ['Dropdown', 'Textarea', 'Date Added'];
+
   state = {
     additionalAttributes: [],
-    attributes: ['Location', 'Asset tag', 'Date Added'].map((attrPlaceholder) => {
+    attributes: ['Location', 'Asset tag', 'Date Added'].map((attrPlaceholder, i) => {
```
Comment (bahdcoder) [inline src/pages/CreateCategoryPage/index.tsx]: For more understandable code, rename this variable to `index`.

File: src/components/SelectBox/index.tsx [hunk 3]
```diff
@@ -7,11 +7,19 @@ import './SelectBox.scss';
 // interfaces
 import { SelectBoxProps } from './interfaces';
 
-const SelectBox = ({ classes, options, onChange, value, name }: SelectBoxProps) => ((
+const SelectBox = ({ classes, options, onChange, value, name, index, defaultValue }: SelectBoxProps) => ((
```
Comment (kevinejiro) [inline src/components/SelectBox/index.tsx]: For readability each of these `({classes, options, onChange, value, name, index, defaultValue})`  can be placed in a new line.

File: src/components/SelectBox/index.tsx [hunk 4]
```diff
@@ -7,11 +7,19 @@ import './SelectBox.scss';
...
+      name={name} 
+      onChange={onChange}
+      defaultValue={defaultValue}
+      value={value} 
+      className={`select-box__content ${classes}`} 
+      disabled={index ? true : false}
+    >
       {name && <option disabled value={name}>Choose one</option>}
-      {options.map(option => <option value={option.name} key={option.id}>{option.name}</option>)}
+      {options.map(option => <option value={index && option.id === 3 ? 'Date Added' : option.name}
```
Comment (kevinejiro) [inline src/components/SelectBox/index.tsx]: This can be made more readably. Try writing on a new line where possible.

File: src/pages/CreateCategoryPage/index.tsx [hunk 5]
```diff
@@ -114,6 +116,8 @@ class CreateCategoryPage extends React.Component<{}, CreateCategoryPageState> {
                     key={`${this.currentTime()}-${index}`}
                     placeholder={value.placeholder}
                     enableConfigOptions={false}
+                    index={index + 1}
```
Comment (bahdcoder) [inline src/pages/CreateCategoryPage/index.tsx]: Been looking through the code to understand exactly what the responsibility of the `index` prop is. 🤔 

Could you please shed more light? It looks like you do this in all form-control components:

`isDisabled={index ? true : false}`

Why ? Couldn't a `disabled` boolean prop suffice? Why does it have to be the `index` from the parent array ?

Please throw more light on this, thanks.

File: src/components/SelectBox/index.tsx [hunk 6]
```diff
@@ -7,11 +7,19 @@ import './SelectBox.scss';
...
+      onChange={onChange}
+      defaultValue={defaultValue}
+      value={value} 
+      className={`select-box__content ${classes}`} 
+      disabled={index ? true : false}
+    >
       {name && <option disabled value={name}>Choose one</option>}
-      {options.map(option => <option value={option.name} key={option.id}>{option.name}</option>)}
+      {options.map(option => <option value={index && option.id === 3 ? 'Date Added' : option.name} 
+      key={option.id}>{index && option.id === 3 ? 'Date Added' : option.name}</option>)}
```
Comment (bahdcoder) [inline src/components/SelectBox/index.tsx]: This component is supposed to be used globally in the project. Hard coding values like `3`, `Date Added` right here is not scalable. I suggest you find a better way to achieve your objective.

File: src/components/SelectBox/index.tsx (hunk 3 above)
Comment (bahdcoder) [inline src/components/SelectBox/index.tsx]: Also, Please add tests for the `defaultValue` functionality in this component.

File: src/pages/CreateCategoryPage/index.tsx (hunk 2 above)
Comment (bahdcoder) [reply src/pages/CreateCategoryPage/index.tsx]: For the attributes, according to your task on the PT board, the `Date Added` should actually be `Date`. It references the attribute type.

File: src/components/InputBox/index.tsx (hunk 1 above)
Comment (Iyke09) [reply src/components/InputBox/index.tsx]: will do

File: src/pages/CreateCategoryPage/index.tsx (hunk 2 above)
Comment (Iyke09) [reply src/pages/CreateCategoryPage/index.tsx]: yea, it says date but the reference image from the invision app says date added.

File: src/pages/CreateCategoryPage/index.tsx (hunk 5 above)
Comment (Iyke09) [reply src/pages/CreateCategoryPage/index.tsx]: Since that component is used in several places, putting a disabled = true would cause all input fields to be disabled.

File: src/components/SelectBox/index.tsx (hunk 6 above)
Comment (Iyke09) [reply src/components/SelectBox/index.tsx]: Did that cause I had to change the option.name from radio buttons to date-added, as I didn't want to change it from the source since it is used in other places.

File: src/components/InputBox/interfaces.d.ts [hunk 7]
```diff
@@ -1,9 +1,10 @@
 export interface InputBoxProps {
   error?: string;
-  placeholder: string;
+  placeholder?: string;
```
Comment (FelixWambiri) [inline src/components/InputBox/interfaces.d.ts]: I think that @kevinejiro comment applies here too.

File: src/components/Attribute/Attribute.scss [hunk 8]
```diff
@@ -38,11 +38,7 @@
       }
     }
     
-    &__draggable-image {
-      cursor: move;
-    }
-    
-    &__draggable-image:hover {
```
Comment (andela-kakpobome) [inline src/components/Attribute/Attribute.scss]: Why are you deleting this change, this was perfectly fine. As we discussed, the scope of your work is to disable the input fields.

File: src/components/InputBox/index.tsx [hunk 9]
```diff
@@ -7,18 +7,19 @@ import './InputBox.scss';
...
     <input
       type="text"
       className={`
       input-box-group__control ${
         error ? 'input-box-group__control--error' : ''
-      }`}
+      } ${isDisabled ? 'input-box-group__control--disabled' : ''}`}
       onChange={onChange}
-      value={value}
+      defaultValue={value}
```
Comment (andela-kakpobome) [inline src/components/InputBox/index.tsx]: What is the value of this change? You realize that this (using) the HTML node's `value` prop is how we set the value of the app, how else will that happen? What am I missing?

File: src/components/SelectBox/SelectBox.test.tsx [hunk 10]
```diff
@@ -60,6 +60,6 @@ describe('The SelectBox component', () => {
     const wrapper = mount(<SelectBox { ...props } />);
     const select = wrapper.find(`.${props.classes}`);
     expect(select.length).toBe(1);
-    expect(select.props().className).toEqual('select-box__content asset-pane-width');
+    expect(select.props().className).toEqual('select-box__content asset-pane-width ');
```
Comment (andela-kakpobome) [inline src/components/SelectBox/SelectBox.test.tsx]: What is the extra space before the `'` about?

File: src/components/SelectBox/index.tsx [hunk 11]
```diff
@@ -7,11 +7,19 @@ import './SelectBox.scss';
...
 import { SelectBoxProps } from './interfaces';
 
-const SelectBox = ({ classes, options, onChange, value, name }: SelectBoxProps) => ((
+const SelectBox = ({ classes, options, onChange, value, name, isDisabled, defaultValue }: SelectBoxProps) => ((
   <div className="select-box">
-    <select name={name} onChange={onChange} value={value} className={`select-box__content ${classes}`}>
+    <select 
+      name={name} 
+      onChange={onChange}
+      defaultValue={defaultValue}
```
Comment (andela-kakpobome) [inline src/components/SelectBox/index.tsx]: I generally see the value of this `defaultValue`, can expand on what it's doing here?

File: src/components/SelectBox/SelectBox.test.tsx (hunk 10 above)
Comment (Iyke09) [reply src/components/SelectBox/SelectBox.test.tsx]: That is because the isDisabled props is undefined and so returns an empty string. 
${isDisabled ? 'select-box__content--disabled' : ''}

File: src/components/Attribute/Attribute.scss (hunk 8 above)
Comment (Iyke09) [reply src/components/Attribute/Attribute.scss]: fixed this

File: src/pages/CreateCategoryPage/interfaces.d.ts [hunk 12]
```diff
@@ -22,4 +22,5 @@ interface Attributes {
   label?: string;
   inputControl: string;
   isRequired: boolean;
+  defaultSelectValue?: string;
```
Comment (andela-kakpobome) [inline src/pages/CreateCategoryPage/interfaces.d.ts]: Why's this still here?

File: src/pages/CreateCategoryPage/index.tsx [hunk 13]
```diff
@@ -288,6 +290,8 @@ export class CreateCategoryPage extends React.Component<CreateCategoryPageProps,
                     key={`${index}`}
                     placeholder={attribute.placeholder}
                     enableConfigOptions={false}
+                    isDisabled={true}
+                    defaultSelectValue={attribute.defaultSelectValue}
```
Comment (andela-kakpobome) [inline src/pages/CreateCategoryPage/index.tsx]: Why's this still here?

File: src/pages/CreateCategoryPage/index.tsx [hunk 14]
```diff
@@ -33,12 +34,13 @@ export class CreateCategoryPage extends React.Component<CreateCategoryPageProps,
     errors: {
       newCategoryName: '',
     },
-    attributes: ['Asset tag'].map((attrPlaceholder) => {
+    attributes: ['Asset tag'].map((attrPlaceholder, i) => {
```
Comment (andela-kakpobome) [inline src/pages/CreateCategoryPage/index.tsx]: `i` is not a descriptive variable name

File: src/pages/CreateCategoryPage/index.tsx [hunk 15]
```diff
@@ -33,12 +34,13 @@ export class CreateCategoryPage extends React.Component<CreateCategoryPageProps,
...
       return {
         label: '',
         inputControl: 'Textarea',
         isRequired: false,
         placeholder: attrPlaceholder,
+        defaultSelectValue: this.defaultSelectValues[i],
```
Comment (andela-kakpobome) [inline src/pages/CreateCategoryPage/index.tsx]: I don't see the point of this, `inputControl` already handles this, we discussed this multiple times in your pairing session

File: src/components/SelectBox/index.tsx [hunk 16]
```diff
@@ -7,11 +7,25 @@ import './SelectBox.scss';
...
+  name, 
+  isDisabled,
+}: SelectBoxProps) => ((
   <div className="select-box">
-    <select name={name} onChange={onChange} value={value} className={`select-box__content ${classes}`}>
+    <select 
+      name={name} 
+      onChange={onChange}
+      value={value} 
+      className={`select-box__content ${classes} ${isDisabled ? 'select-box__content--disabled' : ''}`}
```
Comment (andela-kakpobome) [inline src/components/SelectBox/index.tsx]: This line is too long, figure out a way to have it shorter

File: src/components/SelectBox/SelectBox.test.tsx (hunk 10 above)
Comment (andela-kakpobome) [reply src/components/SelectBox/SelectBox.test.tsx]: After `'select-box__content asset-pane-width` there's an extra space, that's what I'm referring to

File: src/components/Attribute/Attribute.scss [hunk 17]
```diff
@@ -37,14 +37,15 @@
...
     }
-    
-    &__draggable-image {
+
+    &__draggable-image:hover {
       cursor: move;
     }
     
-    &__draggable-image:hover {
+    &--isDragable {
```
Comment (andela-kakpobome) [inline src/components/Attribute/Attribute.scss]: 41 & 45 are the same rule essentially, why can't you summarize both?

File: src/components/Attribute/Attribute.scss [hunk 18]
```diff
@@ -37,14 +37,15 @@
...
       cursor: move;
     }
+
```
Comment (andela-kakpobome) [inline src/components/Attribute/Attribute.scss]: What's this line for?

File: src/components/Attribute/Attribute.scss (hunk 17 above)
Comment (Iyke09) [reply src/components/Attribute/Attribute.scss]: Done!

File: src/components/Attribute/Attribute.scss (hunk 18 above)
Comment (Iyke09) [reply src/components/Attribute/Attribute.scss]: Done!

File: src/pages/CreateCategoryPage/index.tsx (hunk 14 above)
Comment (Iyke09) [reply src/pages/CreateCategoryPage/index.tsx]: Done!

File: src/components/SelectBox/index.tsx (hunk 16 above)
Comment (Iyke09) [reply src/components/SelectBox/index.tsx]: I couldn't find a way to make it shorter except simply breaking it unto a new line.

File: src/pages/CreateCategoryPage/index.tsx (hunk 15 above)
Comment (Iyke09) [reply src/pages/CreateCategoryPage/index.tsx]: fixed

File: src/pages/CreateCategoryPage/index.tsx (hunk 13 above)
Comment (Iyke09) [reply src/pages/CreateCategoryPage/index.tsx]: fixed!

File: src/pages/CreateCategoryPage/interfaces.d.ts (hunk 12 above)
Comment (Iyke09) [reply src/pages/CreateCategoryPage/interfaces.d.ts]: fixed!

File: src/components/SelectBox/SelectBox.test.tsx (hunk 10 above)
Comment (Iyke09) [reply src/components/SelectBox/SelectBox.test.tsx]: Yes, and that's what am saying. The extra space is because of this line:

 className=
{`select-box__content ${classes} ${isDisabled ? 'select-box__content--disabled' : ''}`} 

since the isDisabled props is undefined in the above, it then returns an empty string and hence the space at the end.

File: src/pages/CreateCategoryPage/index.tsx [hunk 19]
```diff
@@ -0,0 +1,230 @@
+// react libraries
```
Comment (Dammyy) [inline src/pages/CreateCategoryPage/index.tsx]: `CreateCategoryPage` no longer exists on the `develop` branch. Please rebase with develop and move your implementation to the appropriate file.

File: src/pages/CreateCategoryPage/index.tsx [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+// third party libraries
+import { DragDropContext, Draggable, Droppable } from 'react-beautiful-dnd';
+
+class CreateCategoryPage extends React.Component<{}, CreateCategoryPageState> {
+  state = {
+    additionalAttributes: [],
+    attributes: ['Asset tag'].map((attrPlaceholder) => {
+      return {
+        label: '',
+        inputControl: 'Textarea',
```
Comment (Dammyy) [inline src/pages/CreateCategoryPage/index.tsx]: The appropriate inputControl for Asset Tag is `text` , not Textarea.

File: src/components/AddAssetForm/interfaces.d.ts [hunk 21]
```diff
@@ -7,7 +7,7 @@ export interface AddAssetFormProps {
   }) => void;
   categories: [
     {
-      id: string,
+      id: number,
```
Comment (Dammyy) [inline src/components/AddAssetForm/interfaces.d.ts]: Revert this change.

File: src/components/Attribute/fixtures.ts [hunk 22]
```diff
@@ -1,5 +1,5 @@
 export const inputTypes = [
-  { id: '1', name: 'Textarea' },
-  { id: '2', name: 'Dropdown' },
-  { id: '3', name: 'Radio buttons' },
+  { id: 1, name: 'Dropdown' },
+  { id: 2, name: 'Textarea' },
+  { id: 3, name: 'Radio buttons' },
```
Comment (Dammyy) [inline src/components/Attribute/fixtures.ts]: These are the inputControls supported by the backend

`TEXTAREA = "textarea"`
`TEXT = "text"`
`DROPDOWN = "dropdown"`
`CHECKBOX = "checkbox"`
`RADIOBUTTON = "radio button"`
`DATE = "date added"`

The values to the right are what you should have as the `name`. Please correct this

File: src/pages/CreateCategoryPage/index.tsx (hunk 19 above)
Comment (FelixWambiri) [reply src/pages/CreateCategoryPage/index.tsx]: +1

**Comment by kevinejiro** [comment]
Please fix drop in the total coverage.

File: src/components/Attribute/fixtures.ts [hunk 23]
```diff
@@ -1,5 +1,6 @@
 export const inputTypes = [
```
Comment (Dammyy) [inline src/components/Attribute/fixtures.ts]: You have to modify `allowedSelectOptions` on line 207 in attributes/index.tsx to the new values for inputTypes `dropdown and radio button`. If you don't do this, the options input field which appears when adding an additional attribute (dropdown or radio button) which enables us set the select options for the dropdown or radio button would not show

File: src/components/InputBox/index.tsx [hunk 24]
```diff
@@ -7,18 +7,27 @@ import './InputBox.scss';
...
+}: InputBoxProps) => (
+  <div className={
+    error ? 'input-box-group input-box-group--error' : 'input-box-group'}>
     <input
       type="text"
       className={`
       input-box-group__control ${
         error ? 'input-box-group__control--error' : ''
-      }`}
+      } ${isDisabled ? 'input-box-group__control--disabled' : ''}`}
```
Comment (andela-kakpobome) [inline src/components/InputBox/index.tsx]: Move these into a variable
